functions/src/SQR.functions.js

```diff
@@ -0,0 +1,234 @@
+
+        if(original.not_preferred_language) 
+            audioFileStatus = 'Spare';        
+        else if(original.unable_to_play_or_download)
+            audioFileStatus = 'Audio Problem';
+        
+
+        // 1. Add the webform data to a SQR submissions DB path
+        let submission = {
+            fileName: original.audio_file_name,
+            cancellation: {
+                notPreferredLanguage: audioFileStatus === 'Spare',
+                audioProblem: audioFileStatus === 'Audio Problem'
+            },
+            soundQualityRating: original.sound_quality_rating,
+            unwantedParts: original.unwanted_parts,
+            soundIssues: original.sound_issues,
+            duration: {
+                beginning: original.beginning,
+                ending: original.ending,
+            },
+            comments: original.comments,
+            token: original.token,
+            created: original.created, 
+            //  timestamp of the submission creation,
+            // can differ from COMPLETED in case of saving a DRAFT and completing later.
+            completed: original.completed, // timestamp of the submission completion.
+            changed: original.changed, //timestamp of the submission update.
+            devotee: {
+                name: original.name,
+                emailAddress: original.email_address
+            }
+        };
+        db.ref(`/sqr/submissions/${original.serial}`).update(submission);
+
+
+
+        // 2. Update the allotment
+        let allotmentUpdates = { status: audioFileStatus };
+
+        // in case 1 & 2 add the comments to the notes
+        if (audioFileStatus !== 'WIP')
+            allotmentUpdates.notes = original.comments;
+        
+        // if the audio has a problem then REMOVE the devotee from the file allotment
+        if(audioFileStatus === 'audioProblem')
+            allotmentUpdates.devotee = {};
+
+        db.ref(`/sqr/files/${original.list}/${original.audio_file_name}`).update(allotmentUpdates);
+
+
+
+
+        // Coordinator object example { templateid: 3, email:'a@a.a', name: 'Aj' }
+        let coordinator = functions.config().audioseva.coordinator;
+        // 3. Notify the coordinator
+        // 3.1 Get the submitted audio file data
+
+        //  EXTRACTING the list name first from the file_name
+        let list = original.audio_file_name.match(/^[^-]*[^ -]/g)[0];
+        let file_name = original.audio_file_name.slice( list.length + 1, original.audio_file_name.length );
+        
+        db.ref(`/sqr/files/${list}/${file_name}`).once('value')
+        .then(snapshot => {
+            if(snapshot.exists()) {
+                let fileData = snapshot.val();
+
+                // 3.2 Get the devotee's Allotments in ('given' || 'WIP') state
+                // Using the /sqr/files here instead of /sqr/allotmens to be able to check for the 
+                // state of ('given' || 'WIP')
+                db.ref(`/sqr/files`).once('value')
```

Comment:
Again, iterating files might be not a good way. Consider querying.

Reply:
In this one particularly, I'm unable to do it.. the structure is as follows:

**--files
----lists [array]
--------file
----------allotment
------------devotee
--------------emailAddress**

what I want to look for is the **emailAddress**, 
In the security rules I added a new index
```
"sqr": {
      "files": {
        "$list": {
         	".indexOn": ["status", "allotment/devotee/emailAddress"]
        }
      },
}
```
and in the JS functions, I used 
```javascript
db.ref(`/sqr/files`).orderByChild('allotment/devotee/emailAddress').equalTo('a@a.a')
```
The problem is that **list name** is variable, is there a way to automatically let Firebase know that (like in the security rules where **$list** is used)?

Reply:
I cannot understand the problem. Can you rephrase the question?

Reply:
I want to search for the **emailAddress** ... this is located in the following structure:

--files
----lists [array]
--------**file**
----------allotment
------------devotee
--------------emailAddress
if the **`file`** ( above in the data strcture ) had no **`lists`** parent I'd have searched like that
```javascript
db.ref(`/sqr/files`).orderByChild('allotment/devotee/emailAddress').equalTo('a@a.a')
```
of course after adding **`".indexOn":  ["allotment/devotee/emailAddress"]`** at **`/sqr/files`** to the security rules.

the problem here is the existence of the **lists** parent, I don't want to make the index on the 
**`/sqr/files`** .. I wan't it to be on **`/sqr/files/list_name`**
this can be achieved in the security rules by having something like this 
```
"sqr": {
      "files": {
        "$list": {
         	".indexOn": ["allotment/devotee/emailAddress"]
        }
      },
}
```

what I couldn't find online is how can I use this new index in the node.js API .. how can I add something like **`$list`** to the code to automatically skip directly to the files and search them on this index **`allotment/devotee/emailAddress`**
```javascript
db.ref(`/sqr/files`).orderByChild('allotment/devotee/emailAddress').equalTo('a@a.a')
```

Reply:
Why don't you want to add index on the files node?

Reply:
Tried that already but didn't work out... it returns **`null`**
**security rules:**
![g1](https://user-images.githubusercontent.com/1976151/47929487-d6518d00-ded1-11e8-825c-00ec14964ac0.png)


**db entry:**
![g2](https://user-images.githubusercontent.com/1976151/47929526-f97c3c80-ded1-11e8-9cc3-77760f37c57f.png)

**JS:**
![g3](https://user-images.githubusercontent.com/1976151/47929550-07ca5880-ded2-11e8-8b0b-0d5d6d946200.png)

**result:**
![g4](https://user-images.githubusercontent.com/1976151/47929558-0e58d000-ded2-11e8-9cef-b91127e4b6aa.png)

Reply:
Tried that already, but it returns **`null`**

**1. Security Rules**
![g1](https://user-images.githubusercontent.com/1976151/47930078-91c6f100-ded3-11e8-82bd-e711a9413cc0.png)

**2. DB entry**
![g2](https://user-images.githubusercontent.com/1976151/47930081-97243b80-ded3-11e8-9d42-586067741fbc.png)


**3. JS**
![g3](https://user-images.githubusercontent.com/1976151/47930087-9a1f2c00-ded3-11e8-8544-7c491c3ebc1a.png)


**4. Result**
![g4](https://user-images.githubusercontent.com/1976151/47930095-a0150d00-ded3-11e8-8294-764e8ea02947.png)

Reply:
Yes, I see that it is not possible to use wildcards in the `orderBy` function.
We'll sort it out later, please remove this code for now and just pass empty array to the email template.

Reply:
This was just an example for explaining my point ... of course in the original code I add the **email of the devotee**.
What that was originally about is me **iterating** the allotments in all the files, you asked me then to do it by querying ... so I just wanted to show that querying is not possible AFAIK in this partiucular case.